Add _eval_refine method to Boolean

JSS95 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->

Derived from #20723

#### Brief description of what is fixed or changed

`refine` on relational object has returned boolean value if the truth value can be determined. This PR moves the logic to `Boolean._eval_refine` so that this can be applied to other boolean objects such as `Predicate`.

Before this PR:
```python
>>> refine(x > 0, Q.is_true(x > 0))
True
>>> refine(Q.positive(x), Q.positive(x)) # inconsistent
Q.positive(x)
```

With this PR:
```python
>>> refine(Q.positive(x), Q.positive(x))
True
```

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
- core
    - `.refine()` method is moved from `Expr` to `Basic`.
- logic
    - `Boolean._eval_refine()` method is defined. This makes every Boolean object, including `Predicate`, show consistent behavior.
<!-- END RELEASE NOTES -->

asmeurer commented:
We should think about what the scope of `refine` should be. Otherwise it will just become another generic `simplify` function without any clear indication on when it should be used or when something else should instead. If the goal here is just to reduce boolean expressions to True or False given other boolean expressions as assumptions, then `ask` already does that. I suppose it's also possible for refine to simplify a boolean expression without completely reducing it to True or False, but I would personally expect that to be a separate function that simplifies booleans given assumptions.

asmeurer commented on sympy/logic/boolalg.py:
```diff
@@ -179,6 +179,10 @@ def binary_symbols(self):
 
+    def _eval_refine(self, assumptions):
+        from sympy.assumptions import ask
+        return ask(self, assumptions)
```
At least it should return the expression unchanged if `ask` returns None. refine is supposed to simplify the input, but `None` is not a simplification of a boolean expression.

JSS95 replied:
That is already handled by `refine`.

```python
>>> print(ask(Q.even(x), Q.real(x)))
None
>>> refine(Q.even(x), Q.real(x))
Q.even(x)
```

JSS95 commented:
@asmeurer 

`ask` returns `None` if the truth value cannot be determined. `refine` returns the boolean expression in such case.

I am implementing binary relation as predicate in #20722. In that PR, I introduced `rearrange` method to do simplification with assumption but without reducing it to `True` or `False`.

JSS95 commented:
If there is no more review, I think this can be merged.

JSS95 commented:
Any more review, please? If not, I think I can merge this.

oscarbenjamin commented on sympy/assumptions/ask.py:
```diff
@@ -1341,6 +1341,11 @@ def ask(proposition, assumptions=True, context=global_assumptions):
 
+    This function evaluates the proposition to ``True`` or ``False`` if
+    the truth value can be determined. If not, it returns ``None``.
+    It should be discerned from :obj:`refine() <sympy.assumptions.refine.refine>`
```
Can this just be
```
:obj:`~.refine`
```
?

JSS95 replied:
It raises sphinx failure.
```
WARNING: more than one target found for cross-reference 'refine()': sympy.assumptions.refine, sympy.assumptions.refine.refine, sympy.core.basic.Basic.refine, sympy.matrices.common.MatrixCommon.refine
```

oscarbenjamin replied:
What about
```
:py:func:`~.refine`
```

JSS95 replied:
It works. Thanks!

oscarbenjamin commented on sympy/assumptions/ask.py:
```diff
@@ -1382,6 +1386,12 @@ def ask(proposition, assumptions=True, context=global_assumptions):
+    See Also
+    ========
+
+    sympy.assumptions.refine.refine : Simplification using assumptions.
```
Can this just be `refine :`?

JSS95 replied:
That raised sphinx failure. It said that multiple target for cross referencing was found.

JSS95 replied:
Is there any way to overcome this? See Also section does not take `:func:` syntax so I can't figure out how to enhance this.

oscarbenjamin commented on sympy/assumptions/refine.py:
```diff
@@ -26,6 +32,16 @@ def refine(expr, assumptions=True):
     x
 
+    >>> refine(Q.real(x), Q.positive(x))
+    True
```
I find this a bit strange. Why is `refine` being used rather than `ask` here?

In what situation does it make sense for `refine` to be applied to predicates like this?

JSS95 replied:
In `master`, `Relational` is evaluated to boolean by `refine`. While implementing `Q.eq`, I found out that other boolean expressions does not follow this so I made the design consistent.

For practical example, consider this situation.
```python
Or(Q.positive(x), Q.negative(x))
```
Reducing this expression using assumption `Q.nonnegative(x)` will return `Q.positive(x)`. We can't use `ask` here, because it will return `None`. In this situation, we need to use `refine`.

oscarbenjamin commented on sympy/core/basic.py:
```diff
@@ -1674,6 +1674,11 @@ def simplify(self, **kwargs):
+    def refine(self, assumption=True):
+        """See the refine function in sympy.assumptions"""
+        from sympy.assumptions import refine
+        return refine(self, assumption)
```
I think `Basic` already has too many methods. Is this needed?

oscarbenjamin replied:
Oh, I see it's just moved from `Expr`... Is that what this change is?

JSS95 replied:
`refine()` function currently supports basic object so that `refine(x>0, Q.is_true(x>0))` return `True`. I think that justifies allowing `(x>0).refine(Q.is_true(x>0))` so I moved this method from `Expr` to here. `Abs(x).refine(Q.positive(x))` has been already supported, after all.

JSS95 replied:
> Oh, I see it's just moved from Expr... Is that what this change is?
Indeed!

oscarbenjamin commented:
Can you explain a bit more why this needed? The rationale is unclear to me.

JSS95 commented:
@oscarbenjamin 
> Can you explain a bit more why this needed? The rationale is unclear to me.

Currently in `master`, refining `Relational` returns boolean value if the truth value can be determined. I think that is reasonable, but I found that this logic is not applied to other boolean objects such as `Predicate`. This PR ensures the consistent behavior.

Ultimately, I want to do something like this:
```python
>>> Or(Eq(x, 1), Eq(x, -1)).refine(Q.positive(x))
Eq(x, 1)
```

JSS95 commented:
@oscarbenjamin 
> Can you explain a bit more why this needed? The rationale is unclear to me.

Currently in `master`, refining `Relational` returns boolean value if the truth value can be determined. I think that is reasonable, but I found that this logic is not applied to other boolean objects such as `Predicate`. This PR ensures the consistent behavior.

Ultimately, I want to combine this to `Q.eq` which is being introduced in #20723, and do something like this:

```python
>>> Or(Q.eq(x, 1), Q.eq(x, -1)).refine(Q.positive(x))
Eq(x, 1)
```

JSS95 committed:
refactor(logic): Implement _eval_refine to Boolean

Make every Boolean follow Relational's refine rule for consistency.
Move .refine() method from Expr to Basic so that non-Expr classes can
use it.

JSS95 commented:
@asmeurer @oscarbenjamin 
Any more review, please? We need this change to ensure consistent `refine` logic over boolean classes.

JSS95 commented:
I will wait a few more days and merge this.
